### PR TITLE
OAK-9721: Add state information while importing index

### PR DESCRIPTION
**Pull request by tihom88**

### CODE REVIEW

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java [hunk 1]
```diff
@@ -58,7 +60,7 @@
      */
     static final String ASYNC_LANE_SYNC = "sync";
 
-    private final Logger log = LoggerFactory.getLogger(getClass());
+    private static final Logger log = LoggerFactory.getLogger(IndexImporter.class);
```
Comment (thomasmueller) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java:63]: Nit: if it's static, then in theory (according to the coding rules), "log" should be "LOG". As you anyway have to change the other constants, I recommend to change this as well.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java [hunk 2]
```diff
@@ -69,6 +71,9 @@
     private final IndexEditorProvider indexEditorProvider;
     private final AsyncIndexerLock indexerLock;
     private final IndexDefinitionUpdater indexDefinitionUpdater;
+    static final int retries = Integer.parseInt(System.getProperty("oak.index.import.retries", "5"));
+    public static final String indexImportStateKey = "indexImportState";
```
Comment (thomasmueller) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java:75]: but here, I think capitalizing it is the right way: INDEX_IMPORT_STATE_KEY

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java [hunk 3]
```diff
@@ -69,6 +71,9 @@
     private final IndexEditorProvider indexEditorProvider;
     private final AsyncIndexerLock indexerLock;
     private final IndexDefinitionUpdater indexDefinitionUpdater;
+    static final int retries = Integer.parseInt(System.getProperty("oak.index.import.retries", "5"));
```
Comment (thomasmueller) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java:74]: should be RETRIES

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java [hunk 4]
```diff
@@ -84,31 +89,74 @@ public IndexImporter(NodeStore nodeStore, File indexDir, IndexEditorProvider ind
...
+            // this is again an idempotent function
+            runWithRetry(retries, IndexImportState.RELEASE_CHECKPOINT, () -> releaseCheckpoint());
+            log.info("Done with releasing checkpoint");
+
+            // Remove indexImportState property on successful import
+            updateIndexImporterState(null, true);
+            log.info("Done with removing index import state");
+
+        } catch (CommitFailedException | IOException e){
+            try{
```
Comment (thomasmueller) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java:140]: Nit: a space after "try" is missing.

Comment (thomasmueller) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java:140]: I think the first thing we should do here is log the error.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java [hunk 5]
```diff
@@ -327,4 +408,25 @@ public String toString() {
...
+        int count = 0;
+        while (count < maxRetries) {
+            log.info("IndexImporterStepExecutor:{} ,count:{}", indexImportState, count);
+            try {
+                step.execute();
+                break;
+            } catch (CommitFailedException | IOException e) {
+                log.info("IndexImporterStepExecutor:{} fail count: {}", indexImportState, count);
+                if (++count >= maxRetries) {
+                    log.info("IndexImporterStepExecutor:{} failed after {} retries", indexImportState, maxRetries);
```
Comment (thomasmueller) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java:425]: H would use log.warn here.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java [hunk 6]
```diff
@@ -327,4 +408,25 @@ public String toString() {
...
+
+    void runWithRetry(int maxRetries, IndexImportState indexImportState, IndexImporterStepExecutor step) throws CommitFailedException, IOException {
+        int count = 0;
+        while (count < maxRetries) {
+            log.info("IndexImporterStepExecutor:{} ,count:{}", indexImportState, count);
+            try {
+                step.execute();
+                break;
+            } catch (CommitFailedException | IOException e) {
+                log.info("IndexImporterStepExecutor:{} fail count: {}", indexImportState, count);
```
Comment (thomasmueller) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporter.java:423]: I would log.warn here, and also log the exception stack trace (that is, append ", e").

File: oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporterTest.java [hunk 7]
```diff
@@ -538,4 +557,153 @@ private void dumpIndexDefinitions(String... indexPaths) throws IOException, Comm
         printer.print(pw, Format.JSON, false);
         Files.write(sw.toString(), file, UTF_8);
     }
+
+    private String importData_IncrementalUpdate_Before_Setup_Method() throws IOException, CommitFailedException {
```
Comment (thomasmueller) [inline oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporterTest.java:561]: That should be importDataIncrementalUpdateBeforeSetup right?

File: oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporterTest.java [hunk 8]
```diff
@@ -538,4 +557,153 @@ private void dumpIndexDefinitions(String... indexPaths) throws IOException, Comm
...
+            @Override
+            public String getType() {
+                return "property";
+            }
+        };
+        return importerProvider;
+    }
+
+    @Test
+    public void importData_IncrementalUpdate_Test_Failure_AT_SWITCHLANE_State() throws Exception {
```
Comment (thomasmueller) [inline oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporterTest.java:613]: This is a mix of camel case, underscore, caps...
importDataIncrementalUpdateTestFailureAtSwitchLaneState

File: oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporterTest.java [hunk 9]
```diff
@@ -538,4 +557,153 @@ private void dumpIndexDefinitions(String... indexPaths) throws IOException, Comm
...
+            assertEquals(of("a", "b", "c", "d"), find(lookup, "foo", "abc", f));
+            assertNotNull(store.retrieve(checkpoint));
+            assertEquals("State matching failed",
+                    IndexImporter.IndexImportState.SWITCHLANE.toString(),
+                    store.getRoot().getChildNode("oak:index").getChildNode("fooIndex").getProperty(IndexImporter.indexImportStateKey).getValue(Type.STRING));
+        }
+    }
+
+    @Test
+    public void importData_IncrementalUpdate_NO_FAILURE() throws Exception {
```
Comment (thomasmueller) [inline oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/index/importer/IndexImporterTest.java:690]: importDataIncrementalUpdateNoFailure

**Review by thomasmueller — CHANGES_REQUESTED** [review]

**Comment by tihom88** [comment]
Raised new PR https://github.com/apache/jackrabbit-oak/pull/527/files. This PR has code conflict with trunk's code.

**Comment by tihom88** [comment]
Raised new PR https://github.com/apache/jackrabbit-oak/pull/527/files. This PR has code conflict with trunk's code.

**Pull request closed without merging**